pkg/authorization/toplevel_org_authorizer.go

```diff
@@ -0,0 +1,155 @@
+func NewTopLevelOrganizationAccessAuthorizer(versionedInformers clientgoinformers.SharedInformerFactory, clusterWorkspaceLister tenancyv1.ClusterWorkspaceLister, delegate authorizer.Authorizer) authorizer.Authorizer {
+	rootKubeInformer := rbacwrapper.FilterInformers(v1alpha1.RootCluster, versionedInformers.Rbac().V1())
+
+	return &topLevelOrgAccessAuthorizer{
+		rootAuthorizer: rbac.New(
+			&rbac.RoleGetter{Lister: rootKubeInformer.Roles().Lister()},
+			&rbac.RoleBindingLister{Lister: rootKubeInformer.RoleBindings().Lister()},
+			&rbac.ClusterRoleGetter{Lister: rootKubeInformer.ClusterRoles().Lister()},
+			&rbac.ClusterRoleBindingLister{Lister: rootKubeInformer.ClusterRoleBindings().Lister()},
+		),
+		clusterWorkspaceLister: clusterWorkspaceLister,
+		delegate:               delegate,
+	}
+}
+
+type topLevelOrgAccessAuthorizer struct {
+	rootAuthorizer         *rbac.RBACAuthorizer
+	clusterWorkspaceLister tenancyv1.ClusterWorkspaceLister
+	delegate               authorizer.Authorizer
+}
+
+func (a *topLevelOrgAccessAuthorizer) Authorize(ctx context.Context, attr authorizer.Attributes) (authorized authorizer.Decision, reason string, err error) {
+	cluster, err := genericapirequest.ValidClusterFrom(ctx)
+	if err != nil || cluster == nil || cluster.Name.Empty() {
+		return authorizer.DecisionNoOpinion, "workspace access not permitted", err
+	}
+
+	workspaceAccessNotPermittedReason := fmt.Sprintf("%q workspace access not permitted", cluster.Name)
+	if !cluster.Name.HasPrefix(v1alpha1.RootCluster) {
+		// nobody other than system:masters (excluded from authz) has access to workspaces not based in root
+		return authorizer.DecisionNoOpinion, workspaceAccessNotPermittedReason, nil
+	}
+
+	// everybody authenticated has access to the root workspace
+	if cluster.Name == v1alpha1.RootCluster {
+		if sets.NewString(attr.GetUser().GetGroups()...).Has("system:authenticated") {
+			return a.delegate.Authorize(ctx, attr)
+		}
+		return authorizer.DecisionNoOpinion, workspaceAccessNotPermittedReason, nil
+	}
+
+	// get org in the root
+	requestTopLevelOrg, ok := topLevelOrg(cluster.Name)
+	if !ok {
+		return authorizer.DecisionNoOpinion, workspaceAccessNotPermittedReason, nil
+	}
+
+	// check the org in the root exists
+	if _, err := a.clusterWorkspaceLister.Get(clusters.ToClusterAwareKey(v1alpha1.RootCluster, requestTopLevelOrg)); err != nil {
+		if errors.IsNotFound(err) {
+			return authorizer.DecisionDeny, workspaceAccessNotPermittedReason, nil
+		}
+		return authorizer.DecisionNoOpinion, workspaceAccessNotPermittedReason, err
+	}
+
+	if subjectCluster := attr.GetUser().GetExtra()[authserviceaccount.ClusterNameKey]; len(subjectCluster) > 0 {
+		// service account will automatically get access to its top-level org
+		subjectTopLevelOrg, ok := topLevelOrg(logicalcluster.New(subjectCluster[0]))
+		if !ok || subjectTopLevelOrg != requestTopLevelOrg {
+			return authorizer.DecisionNoOpinion, workspaceAccessNotPermittedReason, nil
```

Comment:
Does this make it so service accounts are never allowed to access other orgs and any workspaces in other orgs?